### PR TITLE
fix silent installer mode when a previously installed Notepad++ is running

### LINKED ISSUE #10189: Silent Update does not update application if open
### Description of the Issue
When deploying a new version on top an existing one where notepad++ is opened, the notepad++.exe is not updated but the Add/Remove App receives the new version stamp.

### Steps to Reproduce the Issue
1. Install Notepad++ 8.1 using /S
2. Open Notepad++ 8.1
3. Install Notepad++ 8.1.1 /S

### Expected Behavior
Its expected that upon installation open files are checked and the installation does not procceed in case the main application is open

or

The installer force closes the application and re-opens upon installing the new version.
### Actual Behavior
All files but notepad++.exe and nppshell_06.dll got overwritten with the latest version and the Add/Remove Programs is updated showing the version installed as 8.1.1

### PR DESCRIPTION
**Pull request by xomx**
Fix #10189 #10277 https://github.com/microsoft/winget-pkgs/issues/22514 #14236 partially #8514

In the future, this fix should be expanded to include also a new installer option (like "/closeRunningNpp"), which then ensures automatic termination of a previously installed & running Notepad++.

When Notepad++ installer with the '/S' optional param is used while a previously installed Notepad++ instance is running, it does not update the locked main 'notepad++.exe' binary, but the installer still reports success (exit code == 0, regkeys actualized according to the installer version used...).

Moreover - because the installation/upgrade then looks like as being done, some automatic update services like winget cannot be used until a next N++ version is released (winget then erroneously thinks that the current version is already installed...).

There is not such problem when we run that N++ installer in its interactive mode, as it shows the msg "Cannot continue the installation: Notepad++ is running. Please close Notepad++, then click 'Retry'".

### CODE REVIEW

**Comment by Yaron10** [comment]
@xomx,

Thank you for your work. 👍 
I've never used the installer. :)

**Comment by xomx** [comment]
@Yaron10 
I also mostly use the portable N++. But sometimes I also move in an environment restricted by domain policies, limited user rights, etc. So I understand the request of sysadmins for a silent and unattended installation/upgrade.

**Comment by Yaron10** [comment]
@xomx,

I just wanted to thank you for this PR and your other recent PRs.
I've learned it's better not to interfere in PRs related to features I don't normally use.
(That's why I added "I've never used the installer").

**Comment by donho** [comment]
@xomx 
So the current fix is set error code when there's an instance of Notepad++ open, and installer in silent mode. Do you confirm?

File: PowerEditor/installer/nppSetup.nsi [hunk 1]
```diff
@@ -150,6 +150,16 @@ Function .onInit
 	;
 	; --- PATCH END ---
 
+	; handle the possible Silent Mode (/S) & already running Notepad++ (without this an incorrect partial installation is possible)
+	IfSilent 0 notInSilentMode
+	System::Call 'kernel32::OpenMutex(i 0x100000, b 0, t "nppInstance") i .R0'
+	IntCmp $R0 0 nppNotRunning
+	System::Call 'kernel32::CloseHandle(i $R0)' ; a Notepad++ instance is running, tidy-up the opened mutex handle only
```
Comment (donho) [inline PowerEditor/installer/nppSetup.nsi:157]: In the case of `nppNotRunning`, is there no mutex handle ?

Comment (xomx) [reply PowerEditor/installer/nppSetup.nsi:157]: Yes. 
In such case the OpenMutex returns NULL (and possible GetLastError() then shows ERROR_FILE_NOT_FOUND).

**Comment by xomx** [comment]
@donho 
The 3 open issues mentioned above can be also closed.

**Pull request closed without merging**